src/nifs/tests.rs

```diff
@@ -0,0 +1,695 @@
+/// (1) each of the two individual witness-instance pairs satisfy custom gates relation as well
+/// as copy constrains relation (i.e. permutation relation)
+/// (2) the first folded witness-instance pair satisfies the relaxed polynomial relation and
+/// copy constrains relation
+/// (3) the second folded witness-instance pair satisfies the relaxed polynomial relation  and
+/// copy constrains relation
+fn fold_instances<
+    C: CurveAffine<ScalarExt = F1, Base = F2>,
+    F1: PrimeField,
+    F2: PrimeField + FromUniformBytes<64>,
+>(
+    ck: &CommitmentKey<C>,
+    td1: &TableData<F1>,
+    td2: &TableData<F1>,
+) -> Result<(), NIFSError> {
+    const T: usize = 3;
+    const RATE: usize = 2;
+    const R_F: usize = 4;
+    const R_P: usize = 3;
+
+    let S = td1.plonk_structure(ck);
+    let mut f_U =
+        RelaxedPlonkInstance::new(td1.instance.len(), S.num_challenges, S.round_sizes.len());
+    let mut f_W = RelaxedPlonkWitness::new(td1.k, &S.round_sizes);
+    let mut ro_nark_prover = create_ro::<C, F2, T, RATE, R_F, R_P>();
+    let mut ro_nark_prepare = create_ro::<C, F2, T, RATE, R_F, R_P>();
+    let mut ro_nark_verifier = create_ro::<C, F2, T, RATE, R_F, R_P>();
+    let mut ro_nark_decider = create_ro::<C, F2, T, RATE, R_F, R_P>();
+    let (U1, W1) = td1
+        .run_sps_protocol(ck, &mut ro_nark_prepare, S.num_challenges)
+        .unwrap();
+    let res = S.is_sat(ck, &mut ro_nark_decider, &U1, &W1);
+    assert!(res.is_ok());
+
+    let mut ro_acc_prover = create_ro::<C, F2, T, RATE, R_F, R_P>();
+    let mut ro_acc_verifier = create_ro::<C, F2, T, RATE, R_F, R_P>();
+    let (nifs, (_U, W)) =
+        NIFS::prove(ck, &mut ro_nark_prover, &mut ro_acc_prover, td1, &f_U, &f_W)?;
+    let U = nifs
+        .verify(&mut ro_nark_verifier, &mut ro_acc_verifier, &S, f_U, U1)
+        .unwrap();
+    assert_eq!(U, _U);
+
+    f_U = U;
+    f_W = W;
+    let res = S.is_sat_relaxed(ck, &f_U, &f_W);
+    assert!(res.is_ok());
+    //    let perm_res = S.is_sat_perm(&f_U, &f_W);
+    //    assert!(perm_res.is_ok());
+
+    let (U1, W1) = td2
+        .run_sps_protocol(ck, &mut ro_nark_prepare, S.num_challenges)
+        .unwrap();
+    let res = S.is_sat(ck, &mut ro_nark_decider, &U1, &W1);
+    assert!(res.is_ok());
+
+    let (nifs, (_U, _W)) =
+        NIFS::prove(ck, &mut ro_nark_prover, &mut ro_acc_prover, td2, &f_U, &f_W)?;
+    let U = nifs
+        .verify(&mut ro_nark_verifier, &mut ro_acc_verifier, &S, f_U, U1)
+        .unwrap();
+    assert_eq!(U, _U);
+
+    f_U = _U;
+    f_W = _W;
+    let res = S.is_sat_relaxed(ck, &f_U, &f_W);
+    assert!(res.is_ok());
+    //    let perm_res = S.is_sat_perm(&f_U, &f_W);
+    //    assert!(perm_res.is_ok());
+    Ok(())
+}
+
+/// calculate smallest w such that 2^w >= n*(2^K)
+fn smallest_power(n: usize, K: u32) -> usize {
+    let n_f64 = n as f64;
+    let mul_res = n_f64 * (2f64.powi(K as i32));
+    let log_result = mul_res.log2().ceil();
+    log_result as usize
+}
```

Comment:
let's leave it for now because this is in test module and we just use a small number of rows to test. If it is used in application, we can add error handler.